oci/layer/utils_unix.go

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+ * umoci: Umoci Modifies Open Containers' Images
+ * Copyright (C) 2016-2025 SUSE LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package layer
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"slices"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/opencontainers/umoci/pkg/fseval"
+)
+
+type overlayWhiteoutType string
+
+const (
+	overlayWhiteoutPlain  overlayWhiteoutType = "plain whiteout"
+	overlayWhiteoutOpaque overlayWhiteoutType = "opaque whiteout"
+)
+
+// isOverlayWhiteout returns true if the FileInfo represents an overlayfs style
+// whiteout (i.e. mknod c 0 0) and false otherwise.
+func isOverlayWhiteout(path string, fsEval fseval.FsEval) (overlayWhiteoutType, bool, error) {
+	stat, err := fsEval.Lstatx(path)
+	if err != nil {
+		return "", false, err
+	}
+
+	switch stat.Mode & unix.S_IFMT {
+	case unix.S_IFCHR:
+		// classic char 0:0 style whiteouts
+		if stat.Rdev == 0 {
+			return overlayWhiteoutPlain, true, nil
+		}
+	case unix.S_IFDIR:
+		// opaque whiteouts
+		val, err := fsEval.Lgetxattr(path, "trusted.overlay.opaque")
+		if err != nil {
+			// If we are missing privileges to read the xattr (ENODATA) or the
```

Comment:
Is ignoring ENODATA safe? If someone has marked an `/etc/foo.d` directory as a whiteout, lower layers might be able to inject config that higher layers tried to delete. Or am I misunderstanding?

Reply:
My understanding of the usecase is that you wouldn't give untrusted user access to the raw overlayfs-extracted lowerdir, you would give them access to a merged overlayfs right? You can't set those xattrs through a merged overlayfs AFAIK.

(Also for the moment it's `trusted.overlay.opaque` so an unprivileged user couldn't write to it anyway. But this is the concern I had with enabling `user.overlay.*` support by default without making it a config option.)

Reply:
> My understanding of the usecase is that you wouldn't give untrusted user access to the raw overlayfs-extracted lowerdir, you would give them access to a merged overlayfs right?

Maybe, but in stacker's case it will extract them with the unprivileged uid/gid, so it's not impossible. Though I agree, it requires a container escape, and at that point maybe it doesn't matter.

Reply:
I think even in that case you'd be mounting the overlayfs with `-o userxattr` and whatever untrusted program running atop the rootfs won't be able to set the xattrs either.

For an attacker to access the lowerdirs I think you'd need a container escape, as you said.

Reply:
FWIW, I would prefer it if we could avoid this by just checking `llistxattr` (like with `whiteout`) but sadly that's not how that xattr works...

But I also think that `ENODATA` here is ultimately fine because it just means that either umoci isn't privileged enough (we need `userxattr` support to fix this -- #576) or the rootfs has an empty xattr (which could happen for a variety of reasons, and an attacker that can do that can do plenty of other bad things imho).